engine/class_modules/monk/sc_monk.cpp

```diff
@@ -3570,23 +3619,37 @@ struct breath_of_fire_dot_t : public monk_spell_t
   double composite_persistent_multiplier( const action_state_t *state ) const override
   {
     double cpm = monk_spell_t::composite_persistent_multiplier( state );
-    if ( blackout_combo )
+
+    const state_t* s = cast_state( state );
+
+    if ( s->blackout_combo )
       cpm *= 1.0 + p()->buff.blackout_combo->data().effectN( 5 ).percent();
+
     return cpm;
   }
 
-  void execute() override
+  action_state_t* new_state() override
   {
-    // blackout combo buffs only one of the breath of fire dot applications from
-    // a single cast
-    monk_spell_t::execute();
-    blackout_combo = p()->buff.blackout_combo->up();
-    p()->buff.blackout_combo->expire();
+    return new state_t( this, target );
+  }
+
+  state_t* cast_state( action_state_t* s )
+  {
+    return static_cast<state_t*>( s );
+  }
+
+  const state_t* cast_state( const action_state_t* s ) const
+  {
+    return static_cast<const state_t*>( s );
   }
```

Comment:
I prefer directly using `debug_cast` as `cast_state` may be conflated to "a state object/ptr for the cast of the spell" which doesn't exist.

Also `using custom_state_t` is preferable as it makes the fact that the state is non-default more clear, although I'm not entirely stuck on that one.

Reply:
I definitely noticed the conflated words, but truth be told I was just copying what other modules were doing. I like the assertions provided by `debug_cast` - I'll start using that in the future.